src/libmongoc/src/mongoc/mongoc-scram.c

```diff
@@ -1150,4 +1229,176 @@ _mongoc_sasl_prep (const char *in_utf8, int in_utf8_len, bson_error_t *err)
+   case 3:
+      // Four options, separated by ||
+      return (_mongoc_char_between_chars (*c, 0xE0, 0xE0) &&
+              _mongoc_char_between_chars (c[1], 0xA0, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xE1, 0xEC) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xED, 0xED) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0x9F) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xEE, 0xEF) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF));
+   case 4:
+      // Three options, separated by ||
+      return (_mongoc_char_between_chars (*c, 0xF0, 0xF0) &&
+              _mongoc_char_between_chars (c[1], 0x90, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[3], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xF1, 0xF3) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[3], 0x80, 0xBF)) ||
+             (_mongoc_char_between_chars (*c, 0xF4, 0xF4) &&
+              _mongoc_char_between_chars (c[1], 0x80, 0x8F) &&
+              _mongoc_char_between_chars (c[2], 0x80, 0xBF) &&
+              _mongoc_char_between_chars (c[3], 0x80, 0xBF));
+   default:
+      return true;
+   }
+}
+
+
+bool
+_mongoc_char_between_chars (const char c, const char lower, const char upper)
+{
+   return (c >= lower && c <= upper);
+}
+
+bool
+_mongoc_is_code_in_table (unsigned int code,
+                          const unsigned int *table,
+                          int size)
+{
+   // all tables have size / 2 ranges
+   for (int i = 0; i < size; i += 2) {
+      if (code >= table[i] && code <= table[i + 1])
+         return true;
+   }
+
+   return false;
+}
+
+unsigned int
+_mongoc_utf8_to_unicode (const char *c, int length)
+{
+   switch (length) {
+   case 1:
+      return (unsigned int) c[0];
+   case 2:
+      return (unsigned int) (((c[0] & 0x1f) << 6) | (c[1] & 0x3f));
+   case 3:
+      return (unsigned int) (((c[0] & 0x0f) << 12) | ((c[1] & 0x3f) << 6) |
+                             (c[2] & 0x3f));
+   case 4:
+      return (unsigned int) (((c[0] & 0x07) << 18) | ((c[1] & 0x3f) << 12) |
+                             ((c[2] & 0x3f) << 6) | (c[3] & 0x3f));
+   default:
+      return 0;
+   }
+}
+
+int
+_mongoc_unicode_to_utf8 (unsigned int c, char *out)
+{
+   if (c <= 0x7F) {
+      // Plain ASCII
+      out[0] = (char) c;
+      return 1;
+   } else if (c <= 0x07FF) {
+      // 2-byte unicode
+      out[0] = (char) (((c >> 6) & 0x1F) | 0xC0);
```

Comment:
Recommend enabling `-Wconversion` and `-Wsign-conversion` warnings to validate type correctness in bitwise operations on integers with size less than `int`. e.g. this line of code (assuming `uint8_t` consistency as suggested above) may need to look like this:
```c
if (c <= 0x07FFu) {
   ptr[0] = (uint8_t) (((c >> 6) & UINT8_C (0x1F)) | UINT8_C (0xC0));
}
```